lib/nss.c

```diff
@@ -0,0 +1,150 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <dlfcn.h>
+#include <stdbool.h>
+#include <string.h>
+#include <strings.h>
+#include <ctype.h>
+#include "prototypes.h"
+#include "../libsubid/subid.h"
+
+#define NSSWITCH "/etc/nsswitch.conf"
+
+// NSS plugin handling for subids
+// If nsswitch has a line like
+//    subid: sssd
+// then sssd will be consulted for subids.  Unlike normal NSS dbs,
+// only one db is supported at a time.  That's open to debate, but
+// the subids are a pretty limited resource, and local files seem
+// bound to step on any other allocations leading to insecure
+// conditions.
+static bool nss_initialized;
```

Comment:
Btw, probably using `atomic_bool` could be an easy way to achieve thread safety.

Imo, this type and moving `nss_initialized = true;` right after `if (nss_initialized)` check in `nss_init()` could do the trick.

Reply:
But then another thread may assume that nss_init() has completed because nss_initialized == true, while the first thread is still initializing.

Unless I'm missing something.

Reply:
two atomic_flags might work.

Reply:
You are right.